### PR TITLE
Authserver cleanup phase 2 - Split SRP6 into its own file

### PR DESCRIPTION
**Pull request by Treeston**
_(This is part 2 of 5 in the deprecation process of `sha_pass_hash`, as outlined in #25157.)_

The only maintainable crypto is readable crypto.
Our crypto is readable now.

### CODE REVIEW

File: src/common/Common.h [hunk 1]
```diff
@@ -121,4 +122,7 @@ TC_COMMON_API LocaleConstant GetLocaleByName(std::string const& name);
 
 #define MAX_QUERY_LEN 32*1024
 
+constexpr size_t SESSION_KEY_LENGTH = 40;
+using SessionKey = std::array<uint8, SESSION_KEY_LENGTH>;
```
Comment (Treeston) [inline src/common/Common.h:126]: I think this should probably be somewhere else, though I'm not entirely sure _where_.

File: src/server/authserver/Server/AuthSession.cpp [hunk 2]
```diff
@@ -74,8 +74,8 @@ static_assert(sizeof(sAuthLogonChallenge_C) == (1 + 1 + 2 + 4 + 1 + 1 + 1 + 2 +
 typedef struct AUTH_LOGON_PROOF_C
 {
     uint8   cmd;
-    uint8   A[32];
-    Trinity::Crypto::SHA1::Digest M1, crc_hash;
+    Trinity::Crypto::SRP6::EphemeralKey A;
+    Trinity::Crypto::SHA1::Digest clientM, crc_hash;
```
Comment (Carbenium) [inline src/server/authserver/Server/AuthSession.cpp:78]:
```suggestion
    Trinity::Crypto::SHA1::Digest clientM;
    Trinity::Crypto::SHA1::Digest clientM, crc_hash;
```

File: src/common/Cryptography/SRP6.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            static constexpr size_t VERIFIER_LENGTH = 32;
+            using Verifier = std::array<uint8, VERIFIER_LENGTH>;
+            static constexpr size_t EPHEMERAL_KEY_LENGTH = 32;
+            using EphemeralKey = std::array<uint8, EPHEMERAL_KEY_LENGTH>;
+
+            static const std::array<uint8, 1> g;
+            static const std::array<uint8, 32> N;
+
+            // this is the old sha_pass_hash hack
+            [[deprecated("No. Bad. You shouldn't even HAVE that. Stop right now and use MakeRegistrationData instead.")]]
```
Comment (Warpten) [inline src/common/Cryptography/SRP6.h:44]: I'm moderately fine with quips in commit messages but please avoid in strings possibly shown by the compiler.

Comment (Shauren) [reply src/common/Cryptography/SRP6.h:44]: I hate them in commit messages, in compiler output and comments. Remove ALL OF THAT

File: src/server/authserver/Server/AuthSession.h [hunk 4]
```diff
@@ -92,14 +93,13 @@ class AuthSession : public Socket<AuthSession>
...
-    BigNumber b, B;
-    std::array<uint8, 40> sessionKey;
+    std::optional<Trinity::Crypto::SRP6> _srp6;
+    SessionKey _sessionKey;
     std::array<uint8, 16> _reconnectProof;
 
     AuthStatus _status;
     AccountInfo _accountInfo;
-    Optional<std::vector<uint8>> _totpSecret;
+    std::optional<std::vector<uint8>> _totpSecret;
```
Comment (Shauren) [inline src/server/authserver/Server/AuthSession.h:102]: As long as Optional.h header exists, don't use std::optional directly

File: src/server/authserver/Server/AuthSession.cpp [hunk 5]
```diff
@@ -415,32 +406,46 @@ void AuthSession::LogonChallengeCallback(PreparedQueryResult result)
     TC_LOG_DEBUG("network", "database authentication values: v='%s' s='%s'", databaseV.c_str(), databaseS.c_str());
 
     // multiply with 2 since bytes are stored as hexstring
-    if (databaseV.size() != size_t(BufferSizes::SRP_6_V) * 2 || databaseS.size() != size_t(BufferSizes::SRP_6_S) * 2)
-        SetVSFields(rI);
+    if (databaseV.empty() || databaseS.empty())
+    {
+
+#ifndef _MSC_VER
```
Comment (Shauren) [inline src/server/authserver/Server/AuthSession.cpp:412]: Don't use _MSC_VER directly anywhere, we have `TRINITY_COMPILER` defined in CompilerDefs.h

Comment (Shauren) [reply src/server/authserver/Server/AuthSession.cpp:412]: Actually, dont mark the function are deprecated using the attribute then suppress it all over the place... people that use them in their custom code probably don't care about warnings so all you are doing is making our code ugly. Put it in function name `MakeRegistrationDataFromHash_DEPRECATED`

File: src/common/Common.h (hunk 1 above)
Comment (Shauren) [reply src/common/Common.h:126]: AuthCrypt.h

**Comment by aleigood** [comment]
Compile error in Linux：
```
/opt/tc/src/src/server/game/Server/WorldSocket.cpp:248:16: error: declaration of ‘SessionKey AccountInfo::SessionKey’ [-fpermissive]
     SessionKey SessionKey;
                ^~~~~~~~~~
In file included from /opt/tc/src/src/server/game/Server/WorldSession.h:27,
                 from /opt/tc/src/src/server/game/PrecompiledHeaders/gamePCH.h:32,
                 from /opt/tc/src/build/src/server/game/CMakeFiles/game.dir/cmake_pch.hxx:5,
                 from <command-line>:
/opt/tc/src/src/common/Cryptography/Authentication/AuthDefines.h:25:57: error: changes meaning of ‘SessionKey’ from ‘using SessionKey = struct std::array<unsigned char, 40>’ [-fpermissive]
 using SessionKey = std::array<uint8, SESSION_KEY_LENGTH>;
                                                         ^
make[2]: *** [src/server/game/CMakeFiles/game.dir/build.make:3079: src/server/game/CMakeFiles/game.dir/Server/WorldSocket.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1220: src/server/game/CMakeFiles/game.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....

```

**Comment by Treeston** [comment]
2059c32 @aleigood

**Pull request merged**